clang/lib/Sema/SemaExpr.cpp

```diff
@@ -12681,6 +12681,40 @@ QualType Sema::CheckAddressOfOperand(ExprResult &OrigOp, SourceLocation OpLoc) {
     if (isa<CXXDestructorDecl>(MD))
       Diag(OpLoc, diag::err_typecheck_addrof_dtor) << op->getSourceRange();
 
+    if (getASTContext().getTargetInfo().getTriple().getArch() ==
+            llvm::Triple::aarch64 &&
+        getLangOpts().PointerAuthCalls && MD->isVirtual()) {
+      // Argument and return types must be complete.
+      auto ReturnOrParamTypeIsIncomplete = [&](QualType T,
+                                               SourceLocation DeclRefLoc,
+                                               SourceLocation
+                                                   ReturnArgTypeLoc) {
+        if (RequireCompleteType(DeclRefLoc, T, diag::err_incomplete_type)) {
+          Diag(
+              DeclRefLoc,
+              diag::
+                  note_ptrauth_virtual_member_function_pointer_incomplete_arg_return);
+          Diag(
+              ReturnArgTypeLoc,
+              diag::
+                  note_ptrauth_virtual_member_function_incomplete_arg_return_type)
+              << T;
+          return true;
+        }
+        return false;
+      };
+      QualType RetTy = MD->getReturnType();
+      bool IsIncomplete =
+          !RetTy->isVoidType() &&
+          ReturnOrParamTypeIsIncomplete(
+              RetTy, OpLoc, MD->getReturnTypeSourceRange().getBegin());
+      for (auto *PVD : MD->parameters())
+        IsIncomplete |= ReturnOrParamTypeIsIncomplete(PVD->getType(), OpLoc,
+                                                      PVD->getBeginLoc());
+      if (IsIncomplete)
+        return QualType();
+    }
+
```

Comment:
We might need to trigger instantiation of the `noexcept` clause.

Reply:
Sorry, I didn't understand this. Do you mean there should be a test for a virtual function which has the `noexcept` specifier? I'm not sure how that changes the diagnostic.

Reply:
If there's a class template with a virtual function with a dependent `noexcept` clause, and we form a member pointer to that virtual function, we should make sure we don't have problems, which probably means making sure we instantiate the clause.

Reply:
The check `MPTy->isDependentType` isn't needed to emit the diagnostics in the test case, but I added it since it wasn't clear whether there is a guarantee that `Sema::CheckAddressOfOperand` won't be called when the type is dependent.